### PR TITLE
[CALCITE-5508] Allow no-op versions of DATETIME and TIMESTAMP

### PR DESCRIPTION
**Pull request by wnob**
This is a bit of unfinished work from CALCITE-5508. BigQuery supports calling the `TIMESTAMP()` function with a `TIMESTAMP`-typed argument, and the `DATETIME()` function with a `DATETIME`-typed argument. Both functions should simply return their argument in these cases, which is functionality we get for free from the existing implementations.

The problem is that, without proper argument type checking, Calcite will try to coerce the argument types, which in the case of `DATETIME()` causes it to try casting as `DATE` and eliminating the time component.

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java [hunk 1]
```diff
@@ -5055,4 +5055,24 @@ void checkUserDefinedOrderByOver(NullCollation nullCollation) {
     String sql = "SELECT CAST(CAST(? AS INTEGER) AS CHAR)";
     sql(sql).ok();
   }
+
```
Comment (mihaibudiu) [inline core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java:5058]: there is no documentation added in this PR for these behaviors.
Also, it would be nice to have some negative tests, where type inference rejects the query, and with NULL arguments.

Comment (wnob) [reply core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java:5058]: I'll try to come up with a negative test shortly. In the meantime, I've added documentation for the new overloads, and made some minor corrections to the existing documentation so they more accurately reflect how BigQuery actually handles these functions.

The no-op versions are undocumented by BigQuery, but easily verifiable as working with a simple example query: `SELECT DATETIME(DATETIME '2022-02-02 12:12:12')`

File: babel/src/test/resources/sql/big-query.iq [hunk 2]
```diff
@@ -2708,6 +2708,18 @@ SELECT TIMESTAMP(DATE "2008-12-25") AS timestamp_date;
 
 !ok
 
+# This tests the no-op TIMESTAMP and DATETIME functions.
+select TIMESTAMP(TIMESTAMP "2008-01-01 01:03:05") as ts,
```
Comment (tanclary) [inline babel/src/test/resources/sql/big-query.iq:2712]: Maybe using `!type` would be helpful here since the aim is to show they return the same type as they are given.

File: site/_docs/reference.md [hunk 3]
```diff
@@ -2833,12 +2835,13 @@ BigQuery's type system uses confusingly different names for types and functions:
 | b | TIME(timestamp)                                | Extracts the TIME from *timestamp* (a local time; BigQuery's DATETIME type)
 | b | TIME(instant)                                  | Extracts the TIME from *timestampLtz* (an instant; BigQuery's TIMESTAMP type), assuming UTC
 | b | TIME(instant, timeZone)                        | Extracts the time from *timestampLtz* (an instant; BigQuery's TIMESTAMP type), in *timeZone*
+| b | TIMESTAMP(timestampLtz)                        | Returns its argument unchanged (no-op)
```
Comment (tanclary) [inline site/_docs/reference.md:2838]: would `returns *timestampLtz* unchanged (no-op)` fit better with other descriptions?

File: site/_docs/reference.md [hunk 4]
```diff
@@ -2711,9 +2711,11 @@ BigQuery's type system uses confusingly different names for types and functions:
 | p q | DATEADD(timeUnit, integer, datetime)         | Equivalent to `TIMESTAMPADD(timeUnit, integer, datetime)`
 | p q | DATEDIFF(timeUnit, datetime, datetime2)      | Equivalent to `TIMESTAMPDIFF(timeUnit, datetime, datetime2)`
 | q | DATEPART(timeUnit, datetime)                   | Equivalent to `EXTRACT(timeUnit FROM  datetime)`
+| b | DATETIME(timestamp)                            | Returns its argument unchanged (no-op)
```
Comment (tanclary) [inline site/_docs/reference.md:2714]: same comment as below

**Review by tanclary — CHANGES_REQUESTED** [review]
Left a couple of suggestions

File: babel/src/test/resources/sql/big-query.iq (hunk 2 above)
Comment (tanclary) [reply babel/src/test/resources/sql/big-query.iq:2712]: or use `!ok` and `!type` in conjunction somehow
